spec/odp/odp_event_manager_spec.rb

```diff
@@ -0,0 +1,486 @@
+      odp_config.update(nil, nil, nil)
+      event_manager.start!
+
+      event_manager.send_event(**events[0])
+      event_manager.send_event(**events[1])
+      sleep(0.1) until event_manager.instance_variable_get('@event_queue').empty?
+
+      expect(event_manager.instance_variable_get('@current_batch').length).to eq 0
+      expect(spy_logger).not_to have_received(:log).with(Logger::ERROR, anything)
+      expect(spy_logger).to have_received(:log).twice.with(Logger::DEBUG, Optimizely::Helpers::Constants::ODP_LOGS[:ODP_NOT_INTEGRATED])
+      expect(event_manager.running?).to be true
+      event_manager.stop!
+    end
+
+    it 'should log error when queue is full' do
+      allow(SecureRandom).to receive(:uuid).and_return(test_uuid)
+      stub_const('Optimizely::Helpers::Constants::ODP_EVENT_MANAGER', {DEFAULT_QUEUE_CAPACITY: 1})
+      event_manager = Optimizely::OdpEventManager.new(odp_config, logger: spy_logger)
+      allow(event_manager).to receive(:running?).and_return(true)
+
+      event_manager.send_event(**events[0])
+      event_manager.send_event(**events[1])
+      event_manager.flush
+
+      # warning when adding event to full queue
+      expect(spy_logger).to have_received(:log).once.with(Logger::WARN, 'ODP event send failed (queue full).')
+      # error when trying to flush with full queue
+      expect(spy_logger).to have_received(:log).once.with(Logger::ERROR, 'Error flushing ODP event queue.')
+    end
+
+    it 'should log error on exception within thread' do
+      allow(SecureRandom).to receive(:uuid).and_return(test_uuid)
+      event_manager = Optimizely::OdpEventManager.new(odp_config, logger: spy_logger)
+      allow(event_manager).to receive(:add_to_batch).and_raise(StandardError, 'Unexpected error')
+      event_manager.start!
+
+      event_manager.send_event(**events[0])
+      sleep(0.1)
+      event_manager.send_event(**events[0])
+
+      sleep(0.1) until event_manager.instance_variable_get('@event_queue').empty?
+      expect(spy_logger).to have_received(:log).once.with(Logger::ERROR, 'Uncaught exception processing ODP events. Error: Unexpected error')
+      expect(spy_logger).to have_received(:log).once.with(Logger::ERROR, 'ODP event send failed (Queue is down).')
+
+      event_manager.stop!
+    end
+
+    it 'should work with overriden event data' do
+      allow(SecureRandom).to receive(:uuid).and_return(test_uuid)
+      event_manager = Optimizely::OdpEventManager.new(odp_config, logger: spy_logger)
+
+      event = events[0]
+      event[:data][:data_source] = 'my-app'
+      odp_event = Optimizely::OdpEvent.new(**event)
+
+      allow(event_manager.zaius_manager).to receive(:send_odp_events).once.with(api_key, api_host, [odp_event]).and_return(false)
+      event_manager.start!
+
+      event_manager.send_event(**event)
+      event_manager.flush
+      sleep(0.1) until event_manager.instance_variable_get('@event_queue').empty?
+
```

Comment:
Not validating the data valye is overriden?

Reply:
You weren't convinced by the name of the test? 😉